swingstore import should not discard historical metadata, and add repair function

warner opened the pull request:
Previously, the swingstore importer would ignore "historical
metadata": records (with hashes) for transcripts and heap snapshots
that are not strictly necessary to rebuild workers. This was a
mistake: our intention was to always preserve these hashes, so that we
might safely (with integrity) repopulate the corresponding data in the
future, using artifacts from untrusted sources.

This commit rewrites the importer to record *all* metadata records in
the first pass, regardless of whether we want historical data or
not. All of these records will be stubs: missing the actual bundle or
snapshot or transcript items, as if they had been pruned. Then, in the
second pass, we populate those stubs using the matching artifats (or
ignore the historical ones, as configured by the `includeHistorical`
option). A final `assertComplete` pass insists that all the
important (non-historical) records are fully populated.

New tests were added to assert that metadata is preserved regardless
of import mode, and that the `assertComplete` pass really catches
everything.

A new `docs/swingstore.md` was added to describe the data model,
including what it means for records to be pruned.

In addition, a `repairMetadata()` function was added to help with
repairing a swingstore that has been damaged by the problem.

fixes #8025

warner commented:
Additional bugs fixed by this rewrite:

* The [import process](https://github.com/Agoric/agoric-sdk/blob/fb922d3aeddd390919a015a76d5f4e035e6f0240/packages/swing-store/src/swingStore.js#L1017) tolerated `value === null` for `kvStore` and `bundleStore` metadata, but asserted that it was non-`null` for snapStore/transcriptStore metadata. The `getExportData` protocol technically allows `value === null`, and the original export callback would observe this if e.g. a vat was deleted and all its snapshots and transcript spans were deleted. The de-duplicated/flattened version of the export data that makes it all the way to the importer would omit these records completely, rather than having a set followed by a delete, but if we're handling that case for kvStore/bundleStore, then we really should be handling it for the other stores too.
* The ["non-initial transcript spans must have a snapshot to start from" check](https://github.com/Agoric/agoric-sdk/blob/fb922d3aeddd390919a015a76d5f4e035e6f0240/packages/swing-store/src/swingStore.js#L1069-L1071) was too strict: a state-sync snapshot recorded after a vat upgrade but before the first heap snapshot is taken (e.g. immediately after the `startVat` delivery) would leave us with `startPos !== 0` but no current snapshot. That's a perfectly normal export state, but the old consistency check would have rejected it during import.

Regressions:
* The new version lacks the [old version's consistency check](https://github.com/Agoric/agoric-sdk/blob/fb922d3aeddd390919a015a76d5f4e035e6f0240/packages/swing-store/src/swingStore.js#L1073-L1074) which asserted that any `isCurrent === 1` snapshot matches the `startPos` of the is-current transcript span. The better test would be that any is-current transcript span which starts with a `load-snapshot` delivery has a matching snapshot, but I don't want swing-store to know that much detail about the transcript entries.

mhofman commented:
Haven't reviewed yet, but questions based on description

> * The de-duplicated/flattened version of the export data that makes it all the way to the importer would omit these records completely, rather than having a set followed by a delete, but if we're handling that case for kvStore/bundleStore, then we really should be handling it for the other stores too.

The intent to allow `== null` (note I thought we allowed `=== undefined` as well) was indeed to support deletion where the "export data" is a simple stream of the original operations without collapsing. To clarify, the behavior is to also no support this use case for snapshot and transcript metadata ?

> A final `assertComplete` pass insists that all the
> important (non-historical) records are fully populated.

I'm confused as to why this is necessary since I thought the importer figures out the list of required artifacts from the current transcripts and snapshots metadata, and explicitly requires them by name, immediately failing if the artifact is not available.

warner commented:
> The intent to allow `== null` (note I thought we allowed `=== undefined` as well)

yep, both mean deletion, the actual code (written by @FUDCo) is

```js
      if (value == null) {
        // Note '==' rather than '===': any nullish value implies deletion
        kernelStorage.kvStore.delete(subKey);
```

> was indeed to support deletion where the "export data" is a simple stream of the original operations without collapsing. To clarify, the behavior is to now also support this use case for snapshot and transcript metadata ?

I don't expect the importer to be fed with a non-collapsed stream (there's no *good* reason for it to ever observe a deletion), but since the type definition claims that it accepts a `SwingSetExporter`, then I figured it'd be a good idea to make it tolerate precisely that. It didn't add much complexity to the implementation.

If you think it's important to have the importer e.g. throw if it receives a nullish value, I'll change it.

> > A final `assertComplete` pass insists that all the
> > important (non-historical) records are fully populated.
> 
> I'm confused as to why this is necessary since I thought the importer figures out the list of required artifacts from the current transcripts and snapshots metadata, and explicitly requires them by name, immediately failing if the artifact is not available.

Yeah, we talked about that.. when I sat down to the implementation, I realized it was more consistent to require that `getArtifactNames` and the names actually accepted by `getArtifact` be the same list. And thus it was more consistent to have the importer walk `getArtifactNames` and grab everything it offers. In the approach we discussed:
* we'd do a query over each table to find the `isCurrent = 1` records
* we'd compute the artifact name and use `getArtifact()` to import it (ignoring whether or not that name is acknowleged by `getArtifactNames`)
* then after all the operationally-required artifacts are read, we'd walk `getArtifactNames`
* for each name, we'd need to figure out if we processed it already, which means either remembering the list of previously-processed names (so memory usage O(N) in the number of vats), or building a query that checks whether we've already imported that data

In the way I ended up doing it:
* we walk `getArtifactNames` and process everything it offers
* at the end, we check the DB for completeness

Seemed simpler to do the latter, and it also gives us a place to perform more detailed checks like "every current span that doesn't start with `initialize-worker` must have a matching snapshot" (except that actually implementing that check exposed too many details about the transcript record types for swingstore to know about them).

mhofman commented:
> * we'd do a query over each table to find the `isCurrent = 1` records

I believe that query already exists for the export logic?

> * for each name, we'd need to figure out if we processed it already, which means either remembering the list of previously-processed names (so memory usage O(N) in the number of vats), or building a query that checks whether we've already imported that data

Won't we need to do this query anyway when re-populating artifacts (which I understand isn't implemented yet)? In that case you need to know if the artifact offered is already present in the local swing-store before deciding to import it or not?

warner commented on packages/swing-store/src/snapStore.js:
```diff
@@ -81,10 +87,11 @@ export function makeSnapStore(
       PRIMARY KEY (vatID, snapPos),
-      UNIQUE (vatID, inUse),
-      CHECK(compressedSnapshot is not null or inUse is null)
+      UNIQUE (vatID, inUse)
```
Oh rats, I just realized this change (removing the CHECK) means this PR won't work across an upgrade. I needed to remove it to allow the import process to create the record without a snapshot first, then populate the `compressedSnapshot` column a moment later.

I need to think of a way to modify this table at startup, if the check is present, or to use `compressedSnapshot = ""` or `0` or something to represent "not populated".

warner replied:
Or.. maybe this isn't a significant problem. The mainnet chain contains swing-store version 0.9.1. The only existing mainnet DBs were created by one of two pathways:
* `agoric-upgrade-10`, which moved to `mainnet1B` and does an `initSwingStore`
* a state-sync import using `mainnet1B`, which does the old (lossy) `importSwingStore`
both cases will get you a DB with the `CHECK(compressedSnapshot is not null or inUse is null)` constraint.

Let's say the version we're building here is named `0.10.0`, and is deployed to mainnet via `agoric-upgrade-11`. Those two pathways will cause:
* `agoric-upgrade-11` will leave the 0.9.1 schema in place
* state-sync import will create a DB with the 0.10.0 schema (lacking the CHECK)

So it's ok for `importSwingStore` to violate the old CHECK constraint, because that code will only encounter a fresh DB, with the 0.10.0 schema.

The normal runtime `snapStore.saveSnapshot()` function might encounter either schema, so it must obey the CHECK constraint. I didn't change that method, so it's still using the one-statement `sqlSaveSnapshot` operation, which populates `compressedSnapshot` immediately.

We'll wind up with some mainnet instances with one schema, and some with the other.

Either we need to make this code do an ALTER TABLE to remove the CHECK constraint from everybody, so they all wind up with the same schema, or we need to safely tolerate either schema going forward.

Hm.. how should I write a test that makes sure we don't accidentally change `saveSnapshot` in the future? Maybe a test which does `initSwingStore`, then manually doe an `ALTER TABLE` to add the CHECK that just got removed, then run the normal `test-snapstore.js` tests?

mhofman replied:
Please see https://github.com/Agoric/agoric-sdk/pull/7635 as it requires removing the constraint as well.

warner replied:
I'm going to sketch out the process we'd use for schema migration, but not implement it for this release. We'll need to implement it before landing #7635, or before making other changes to the way we use the table, but I'd like to avoid the churn/risk of doing it now (especially because of the more extensive testing we need to build to have confidence in it). See https://github.com/Agoric/agoric-sdk/issues/8089 for details.

warner commented:
> > * we'd do a query over each table to find the `isCurrent = 1` records
> 
> I believe that query already exists for the export logic?

There's a set of `store.getExportRecords()`, and if we set its `includeHistorical` argument to `false`, then they'd only provide the `isCurrent = 1` records. I can't say I really like that argument, though, and for simplicity I'd prefer to remove it (so there'd be no way to accidentally omit integrity-critical hashes during export time).

> > * for each name, we'd need to figure out if we processed it already, which means either remembering the list of previously-processed names (so memory usage O(N) in the number of vats), or building a query that checks whether we've already imported that data
> 
> Won't we need to do this query anyway when re-populating artifacts (which I understand isn't implemented yet)? In that case you need to know if the artifact offered is already present in the local swing-store before deciding to import it or not?

Repopulating artifacts is not implemented yet, correct. I wanted to minimize the changes to the most critical ones. My plan is for the new `repopulate.js` to use the first approach, and change the e.g. `populateSnapshot()` functions introduced in this PR to become tolerant of pre-existing data (which is why they get e.g. `dataProvider` instead of the actual data, so they can choose to ignore it).

mhofman commented on packages/swing-store/src/transcriptStore.js:
```diff
@@ -516,55 +520,109 @@ export function makeTranscriptStore(
+    );
+  }
+
+  const sqlGetSpanMetadataFor = db.prepare(`
+    SELECT hash, isCurrent, incarnation
+      FROM transcriptSpans
+      WHERE vatID = ? AND startPos = ? AND endPos = ?
```
I feel like it'd be more a more versatile query if the `endPos` was selected, and not part of the condition.

warner replied:
Hm, maybe. What I need it for is to see if we have a validation hash for a particular span artifact (in `populateTranscriptSpan`), which has a specific `endPos` in mind. It's also used by `repairTranscriptSpanRecord` (to see if the record already exists).

Changing it to return `endPos` would need an `assert.equal` in both places. In `populateTranscriptSpan`, a mismatch would want to emit the same "no metadata for that thing" error that the zero-row return would provoke. For `repairTranscriptSpanRecord`, I guess it would fare better in the error case where the DB already has a span record for deliveries `4..8`, but the repairing metadata has one for `4..9` (for the same vatID). My version would think that the `4..9` record is new, and try to insert it, which would fail the PRIMARY KEY uniqueness constraint. An explicit check would be able to throw a "metadata mismatch" error.

If "more versatile" is the only improvement, I'm not sure it's worth a change. All the use cases have a specific `endPos` in mind, so it felt more natural to include it in the query. But I'd accept an argument that it would catch more errors or report errors better.

mhofman replied:
I think `populateTranscriptSpan` should skip any transcript artifact that contains less than the number of entries dictated by the local span. It could use artifacts that have more entries than the span by ignoring the extra lines, but only for the current span.
That would support using artifacts that are both newer and older to populate transcript spans

warner replied:
I'll switch this query to return `endPos` instead of taking it.

`populateTranscriptSpan` will throw if `endPos` does not match. I don't think I want it trying to utilize mismatched spans, even if the supplied one is larger than expected (and thus supposedly has all the items it needs, plus more). There's no good reason that they would not match: historical spans have fixed+known bounds, and the current span must not be pruned.

warner replied:
addressed in 4020cac69a004ea274b46158e83675b8e59fd0f7

mhofman commented on packages/swing-store/src/transcriptStore.js:
```diff
@@ -516,55 +520,109 @@ export function makeTranscriptStore(
+
+  const sqlCountPopulatedSpanItems = db.prepare(`
+    SELECT COUNT(*) FROM transcriptItems
+      WHERE vatID = ? AND incarnation = ? AND position >= ? AND position < ?
```
Why do we need to include `incarnation` in the conditions?

warner replied:
Hm, ok `position` is not reset on each new incarnation, so yeah strictly speaking we could omit that from the WHERE clause without affecting the result. I probably forgot about that while writing it.

I'm not inclined to remove it, under the "more specific is probably better" doctrine. But I'll add a comment.

mhofman commented on packages/swing-store/src/repair-metadata.js:
```diff
@@ -0,0 +1,82 @@
+ *
+ * It will not call `exporter.getArtifactNames` or `getArtifacts`.
+ *
+ * At the end of the process, the DB changes will be committed. The
+ * function does not return a SwingStore: most callers will want to
+ * call `openSwingStore` shortly after this repair process.
```
Please return the swing-store and not commit. The importer does its own consistency check on the `blockHeight` saved in the host section before commiting.

mhofman replied:
Thinking about this more, I can live with an implicit commit, as long as only non-current metadata is checked for consistency on repair.

warner replied:
I'll change this to whatever you think is best.

One downside of not doing the commit locally is that it becomes the caller's responsibility, and doing `await repairMetadata(..)` instead of `const { hostStorage } = await repairMetadata(..); await hostStorage.commit()` is a great way to silently discard all the changes (which, since it's merely a repair and not a populate-everything, could easily go unnoticed).

I'm not sure I want to restrict `assertComplete()` to only checking non-current metadata.. in fact the current code only checks *current* data (it asserts that all bundles are populated, all current transcript spans are fully populated, and all current snapshots are populated, so it's looking at more than just metadata). So if the checks are likely to interfere with the cosmic-swingset use case, we should change the behavior.

(I can't think of any way that repair-metadata.js would have opinions on the `blockHeight` field, or the semantics of anything else in kvStore)

mhofman replied:
As discussed, we'll handle the hangover case by checking the block height matches at the cosmos / cosmic-swingset layer. For that however it'd be easier to return a swing-store before commit, and defer commiting to cosmic-swingset

warner replied:
I think I understand the hangover problem, but as you say our defense against corruption is to check the block height before beginning the export-data IAVL migration.

If I understand it correctly, it's not that the caller of `repairMetadata()` needs the swingstore code to refrain from doing a commit, but rather it's that the caller needs to check the blockHeight before moving on to the next step, and to do that, they need a `SwingStore` instance, and it would be more convenient for that code to do:
```js
const swingStore = await repairMetadata(dirPath, exporter);
await swingStore.hostStorage.commit();
const ssHeight = swingStore.hostStorage.kvStore.get('blockHeight');
compare(ssHeight, iavlHeight);
const exporter = makeSwingStoreExporter(dirPath);
// delete old IAVL data, use exporter.getExportData() to repopulate
```

than to do:
```js
await repairMetadata(dirPath, exporter);
const swingStore = openSwingStore(dirPath);
const ssHeight = swingStore.hostStorage.kvStore.get('blockHeight');
compare(ssHeight, iavlHeight);
const exporter = makeSwingStoreExporter(dirPath);
// delete old IAVL data, use exporter.getExportData() to repopulate
```

Writing it out that way, I think having `repairMetadata()` do the commit internally looks cleaner. And requiring the caller to remember to `commit()` feels like a bug farm.

What's your current thinking here?

mhofman replied:
The example above is wrong, I'd basically have done:
```js
const swingStore = await repairMetadata(dirPath, exporter);
const ssHeight = swingStore.hostStorage.kvStore.get('blockHeight');
compare(ssHeight, iavlHeight);
await swingStore.hostStorage.commit();
await swingStore.hostStorage.close();
// return to golang, which will initiate the other steps independently, 
// like delete old IAVL data, and start a new export to repopulate
```

Right now I have to explicitly open the DB instead to do the check, and decided to do it first:
https://github.com/Agoric/agoric-sdk/pull/8143/commits/de0c3e6fd0eeded3797a958be9662a6ab09ebef2

So roughly:
```js
const swingStore = openSwingStore(dirPath);
const ssHeight = swingStore.hostStorage.kvStore.get('blockHeight');
compare(ssHeight, iavlHeight);
await swingStore.hostStorage.close();
await repairMetadata(dirPath, exporter);
// return to golang, which will initiate the other steps independently,
// like delete old IAVL data, and start a new export to repopulate
```

Thinking about this more, the most logical thing to do would be for `repairMetadata` to be a method on an already opened swing-store, so the check can be done first, then kick off the repair, and like any operation mutating an open DB, still require an explicit `commit()`, like so:

```js
const swingStore = openSwingStore(dirPath);
const ssHeight = swingStore.hostStorage.kvStore.get('blockHeight');
compare(ssHeight, iavlHeight);
await swingStore.repairMetadata(exporter);
await swingStore.hostStorage.commit();
await swingStore.hostStorage.close();
// return to golang, which will initiate the other steps independently,
// like delete old IAVL data, and start a new export to repopulate
```

warner replied:
I don't think those `hostStorage.close()` calls are correct: that does `db = null` internally, which means the entire `swingStore` instance is now dead; `close()` is meant to be the very last thing you do before dropping the `swingStore` entirely.

But apart from that, your last example seems good to me: open a swingstore, assert that the height matches the IAVL's value (otherwise the repair might be populating the wrong metadata records, if a snapshot or span was created between the two heights), do the repair, commit, then *don't* `close()`, and continue to use the same `swingStore` instance for everything else the host wants to do.

I'll add `repairMetadata` as a `hostStorage` method, next to the `commit` that the caller must use afterwards (and since `setExportCallback` is sort of related to metadata anyways).

mhofman commented on packages/swing-store/src/repair-metadata.js:
```diff
@@ -0,0 +1,82 @@
+  // and do a completeness check
+  assertComplete(internal, 'operational');
+  await exporter.close();
+  hostStorage.commit();
```
As lint says, missing an `await`

mhofman replied:
yeah it's one of those checks that are too expensive to run.

mhofman commented on packages/swing-store/src/transcriptStore.js:
```diff
@@ -608,6 +609,29 @@ export function makeTranscriptStore(
+        throw Fail`repairTranscriptSpanRecord metadata mismatch: ${existing} vs ${metadata}`;
+      }
+    } else {
+      sqlWriteSpan.run(
```
Yeah here again we're relying on the `(vatID, startPos)` uniqueness. I'd much prefer if the get metadata query returned the `endPos` and we manually checked consistency to return a meaningful error.

warner replied:
addressed in 4020cac69a004ea274b46158e83675b8e59fd0f7

mhofman reviewed (changes requested):
I think this PR regresses on validation of metadata shape. I admit the format is a little confusing (I had argued for less duplication of information), but since we're there, let's verify the consistency as much as possible.

warner commented on packages/swing-store/docs/swingstore.md:
```diff
@@ -0,0 +1,52 @@
+
+Within each incarnation, execution is broken into one or more "spans", with a "current span" and zero or more "historic spans". This breaks up the transcript into corresponding spans.
+
+Each historic span ends with a `save-snapshot` entry which records the creation and saving of an XS heap snapshot. The initial span starts with a `start-worker` entry, while all non-initial spans start with a `load-snapshot` entry.
```
sure, although we don't mark the end of an incarnation with anything special, but it can't hurt to mention that fact

mhofman commented on packages/swing-store/src/repair-metadata.js:
```diff
@@ -0,0 +1,82 @@
+ * all entries to do one of three things:
+ *
+ * 1: kvStore records are ignored (not metadata)
+ * 2: bundle/snapshot/transcript records whose keys already exist will
+ *    be compared against the existing data, and an error thrown if
+ *    they do not match
```
Thinking more about this, can we change this behavior slightly to facilitate handling of handover inconsistency:

if the record is for "current" / "in-use" it's ignored (or at most check that hashes match in the case of bundle & snapshots, but not transcripts)

idempotent add (aka check same if present) only for non current / in-use.

That allows to re-run the repair step even if the swing-store committed at end of upgrade block, but somehow the cosmos DB didn't and crashed before.

warner replied:
ah, hm, lemme think about this

mhofman replied:
After further discussion, let's not do this.

mhofman reviewed (commented):
Further thoughts based on hangover inconsistency handling.

warner commented on packages/swing-store/src/bundleStore.js:
```diff
@@ -62,20 +66,36 @@ export function makeBundleStore(db, ensureTxn, noteExport = () => {}) {
+
+  function populateBundle(bundleID, serialized) {
     ensureTxn();
+    sqlHasBundleRecord.run(bundleID) || Fail`missing ${bundleID}`;
```
In SQL, `UPDATE bundles SET bundle = ? WHERE bundleID = ?` will silently do nothing when WHERE constraint doesn't match any rows. (Just like it will silently corrupt your entire table if you forget the WHERE clause). So the explicit check is actually load-bearing.

mhofman replied:
Ugh, update silently doing nothing is somewhat surprising...

warner commented on packages/swing-store/src/bundleStore.js:
```diff
@@ -62,20 +66,36 @@ export function makeBundleStore(db, ensureTxn, noteExport = () => {}) {
+  function populateBundle(bundleID, serialized) {
     ensureTxn();
+    sqlHasBundleRecord.run(bundleID) || Fail`missing ${bundleID}`;
+    sqlPopulateBundleRecord.run(serialized, bundleID);
```
SQLite has a named parameter thing, but since the `sqlPopulateBundleRecord` definition is so close to the invocation, I didn't bother to use it. https://www.sqlite.org/lang_expr.html#varparam plus https://github.com/WiseLibs/better-sqlite3/blob/master/docs/api.md#binding-parameters , so something like:
```js
  const sqlPopulateBundleRecord = db.prepare(`
    UPDATE bundles SET bundle = $serialized WHERE bundleID = $bundleID
  `);

  ..

    sqlPopulateBundleRecord.run({ bundleID, serialized });
```

.. actually, that looks kinda nice, I'll switch to it.

mhofman replied:
Oh I wish we had done that everywhere! I hate positional params

warner commented on packages/swing-store/src/bundleStore.js:
```diff
@@ -209,23 +268,17 @@ export function makeBundleStore(db, ensureTxn, noteExport = () => {}) {
-    const artifactChunks = exporter.getArtifact(name);
-    const inStream = Readable.from(artifactChunks);
-    const data = await buffer(inStream);
+    const [_tag, bundleID] = name.split('.', 2);
```
Oh, I misread the docs, I mistakenly thought `'foo.bar.baz'.split('.', 2)` would yield `['foo', 'bar.baz']` (too much python still in my head :), but right it would return `['foo', 'bar']` and discard the rest.

I'll change it to recover the previous strictness.

warner commented on packages/swing-store/src/exporter.js:
```diff
@@ -140,7 +140,7 @@ export function makeSwingStoreExporter(dirPath, exportMode = 'current') {
-      return snapStore.exportSnapshot(name, exportHistoricalSnapshots);
+      return snapStore.exportSnapshot(name);
     } else if (type === 'transcript') {
       return transcriptStore.exportSpan(name, exportHistoricalTranscripts);
```
yeah, good question, I'll remove both

warner commented on packages/swing-store/src/importer.js:
```diff
@@ -1,43 +1,36 @@
- *
- * @typedef { import('./exporter').SwingStoreExporter } SwingStoreExporter
- * @typedef { import('./swingStore').SwingStore } SwingStore
- * @typedef {(exporter: SwingStoreExporter) => Promise<SwingStore>} ImportSwingStore
```
Yeah, it wasn't actually used before. The original (in `swingStore.js`) defined `ImportSwingStore` up top, but then the actual `importSwingStore()` definition:

https://github.com/Agoric/agoric-sdk/blob/8baf0aabe3cadeb1cd07fd30641d11baf3e68eaf/packages/swing-store/src/swingStore.js#L970-L976

had it's own local signature, with two additional arguments.

warner commented on packages/swing-store/src/importer.js:
```diff
@@ -47,158 +40,109 @@ export async function importSwingStore(exporter, dirPath = null, options = {}) {
+      if (endPos !== 'current') {
+        internal.snapStore.addSnapshotRecord(JSON.parse(value));
+      }
+      // metadata['snapshot.v1.current'] = 'snapshot.v1.5' , i.e. it
+      // points to the name of the current artifact. We could
+      // conceivably remember this and compare it against the .inUse
+      // property of that record, but it's not worth the effort (we
+      // might encounter the records in either order).
```
Without it, the only way to locate the active snapshot for each vat from the export data is to parse the value of every `snapshot.v1.*` entry, looking for the one with `inUse: 1`, which is O(N) in the number of snapshots.

With it, we can do it in two steps: read the value of `snapshot.v1.current`, then use that value as the second key to look up.

We aren't doing this sort of query against the export data when we do an import, but I wanted to keep the format easy to parse in case some other need came along.

I suppose a valid question is what happens if `snapshot.vNN.current` is missing, or the value is malformed, or the value points to a snapshot record which has `inUse: 0`. And I suppose the answer is that I don't need the importer to also be a thorough correctness checker :). This one record isn't necessary to do an import, so all I really want to do is ignore it.

warner commented on packages/swing-store/src/importer.js:
```diff
@@ -47,158 +40,109 @@ export async function importSwingStore(exporter, dirPath = null, options = {}) {
+      const bundleID = pieces[0];
+      internal.bundleStore.addBundleRecord(bundleID);
+    } else if (tag === 'snapshot') {
+      const [_vatID, endPos] = pieces;
```
Yeah, I suppose.. I do want this whole section to remain short enough to fit on the screen at once, and doing a lot of extra validation distracts from the primary logic ("read tag, dispatch to a store"). 

The export-data *keys* exist to:
* 1: give us a handle, so we can modify or delete the same record later (e.g. when we append to a transcript span, we need to tell the exportCallback to change the previous record's value, increasing `endPos` and changing `hash`)
* 2: allow a sorted dump of the export-data to tell a useful story to a human

The duplication of fields between key and value happens because it would be super-annoying to omit that data from the value. The importing code would have to synthesize/merge a record out of some parsed bits of the key, and the parsed bits of the value. And a human reader of the export-data would have to look at both columns to figure out what was going on. I agree that it would remove the error/mismatch cases ("normal form" usually does), but I suspect it would be a net loss for legibility.

The import process doesn't really need much out of the keys. It needs the tag so it can dispatch the *value* to the right store, but it could do its job with just the tag and the value. I wanted to reflect that truth in this loop.

The bundleID is one exception: the value is pure bundle, and we need the key to hold ID. I could tidy that up by changing `addBundleRecord` to take `(key, value)`, I suppose. Passing `JSON.parse(value)` into the `add*Record` methods makes their function signatures nicer to read, but if I wanted to stick to this "just tag and value" story I guess I could change their signatures to `(key, value)` too.

The `if (!!key.endsWith('.current' ..` check seemed like a good idea, but it also detracts from this story.

So, some options are:
* 1: remove the validation from this loop, pass `(key, value)` to each `add*Record` function
  * this might argue for changing the method names, since we aren't giving them a record, and giving them key+value means they're tied more closely to the import process (specifically tied to the format of export-data), so maybe `importBundleRecord`
* 2: leave it as-is: no more validation, no less
* 3: add more validation: `pieces.length`, parse the value, assert that vatID matches in key and value, assert that startPos or endPos or snapPos matches

I'm inclined to stick with 2, but let me know what you feel about it.

mhofman replied:
I'm actually thinking a plain pass-through would be good.

warner replied:
Ok, I'll rewrite this with option 1: do just enough parsing to figure out which store owns the record, then pass through to e.g. `snapStore.importSnapshotRecord(key, value)`.

warner replied:
pushed, in 821c41dd73622cd090170bfbaad663b55167d8e7

warner commented on packages/swing-store/src/importer.js:
```diff
@@ -47,158 +40,109 @@ export async function importSwingStore(exporter, dirPath = null, options = {}) {
+  // Now take each metadata record and install the stub/pruned entry
+  // into the DB.
+
+  for (const [key, value] of allMetadata.entries()) {
+    const [tag, ...pieces] = key.split('.');
```
I figured that this dispatch-to-a-specific-store function should be as agnostic as possible about the structure of the name. It certainly needs to know the tag (to know where to dispatch it too), and we need to recognize `snapshot.vNN.current` specially (since there's no corresponding metadata record), and, well, ok we also need to spot `transcript.vNN.current` to do that extra bit of validation.

So I wasn't eager to have this function think it knows exactly how many `pieces` should be present. But, I see that I can't be consistent about that, given the other needs.

warner replied:
I changed this to only look at `tag`, and leave all other interpretation of the key and value up to the specific store.

warner commented on packages/swing-store/src/importer.js:
```diff
@@ -47,158 +40,109 @@ export async function importSwingStore(exporter, dirPath = null, options = {}) {
+      internal.transcriptStore.addTranscriptSpanRecord(metadata);
     } else {
-      snapshotInfo.snapPos + 1 === transcriptInfo.startPos ||
-        Fail`current transcript for vat ${q(vatID)} doesn't go with snapshot`;
```
Yeah, I removed it. It worked in the old implementation because we gathered information about each vat as we walked the export data (the `vatArtifacts` Map of `vatInfo = { snapshotKey, transcriptKey }` records). That Map was then used to decide which vats to do anything with, and then populate both the metadata record *and* import the artifact for the current snapshot and/or transcript span. This is the approach that made it too easy to ignore the historical metadata.

In the new implementation, I don't build a table indexed by vats. I add all metadata, then I import all artifacts that the data source provides. So there's no point where I've got both a `snapshotInfo` and a `transcriptInfo` for the same vat.

(also, a weaker argument, I didn't want the import process memory usage to be O(N) in the number of vats, however my decision to de-duplicate/apply-deletions the exportData into `allMetadata` undermines that.. it's actually O(N) in the number of metadata records, which is far larger than the number of vats)

I thought it would be better to add this to the completeness check, than try to track it in the import loops. But the completeness checks are delegated to the individual stores, none of which get to see both transcripts and snapshots at the same time.

So I figured the check wasn't really that important, and too hard to implement, so I left it out. I did leave a TODO in `importer.js > assertComplete` which would do more than the original:

```js
export function assertComplete(internal, level) {
  ..
  internal.bundleStore.assertComplete(level);
  internal.transcriptStore.assertComplete(level);
  internal.snapStore.assertComplete(level);

  // TODO: every isCurrent span that starts with load-snapshot has a
  // matching snapshot (counter-argument: swing-store should not know
  // those details about transcript entries)
}
```

warner commented on packages/swing-store/src/importer.js:
```diff
@@ -47,158 +40,109 @@ export async function importSwingStore(exporter, dirPath = null, options = {}) {
-    );
-    fetchedArtifacts.add(transcriptArtifactName);
-  }
-  const bundleArtifactNames = Array.from(bundleArtifacts.keys()).sort();
```
I've encouraged extra `.sort()` calls as a sort of superstitious defense against nondeterminism surprises. Isolating the influence that one data provider can have over the next step in the processing chain. In this case, the order of `bundleArtifacts.keys()` would match the order of the export-data entries (because of the de-duplication/de-deletion loop earlier), and the `sort` explicitly discards that order. SQLite will insert rows in the same order as `bundleArtifactNames`, and the rows probably have internal ID numbers, which would probably influence a dump if we did a `SELECT * FROM` without any `ORDER BY` clause. So by sorting on the way in, we increase the chances that a simple `sqlite3 swingstore.sqlite .dump` will produce the same order regardless of how the export-data was handled (in particular whether it was de-duped first or not).

Probably not too important, but I figure one of these days, there will be a consensus violation near miss, that we only dodge because some bit of code decided to do a redundant `sort()`.

warner commented on packages/swing-store/src/snapStore.js:
```diff
@@ -81,10 +87,11 @@ export function makeSnapStore(
     )
-  `);
+`);
+  // pruned snapshots will have compressedSnapshot of NULL, and might
+  // also have NULL for uncompressedSize and compressedSize
```
yeah, I think we just treat those columns as best-effort (they're already not included in consensus in any way, and as derivable-from-blob fields we don't include them in the export).. they mostly exist for easy stats queries like "how many bytes are we spending holding snapshots anyways?"

warner commented on packages/swing-store/src/transcriptStore.js:
```diff
@@ -516,55 +520,109 @@ export function makeTranscriptStore(
+    incarnation !== undefined ||
+      Fail`transcript metadata missing incarnation: ${metadata}`;
+
+    sqlWriteSpan.run(
```
sure, I'll add:

```
    // sqlWriteSpan is an INSERT, so the PRIMARY KEY (vatID, position)
    // constraint will catch broken export-data errors like trying to
    // add two different versions of the same span (e.g. one holding
    // items 4..8, a second holding 4..9)
```

warner commented:
I added your fix for the empty span artifacts, and added more tests. I still need to think about the commit question, and  https://github.com/Agoric/agoric-sdk/pull/8075#discussion_r1271712668

gibson042 commented on packages/swing-store/src/snapStore.js:
```diff
@@ -448,62 +482,122 @@ export function makeSnapStore(
    * @returns {Promise<void>}
    */
-  async function importSnapshot(name, exporter, info) {
+  async function populateSnapshot(name, chunkProvider, includeHistorical) {
```
For readability at call sites, I think I'd prefer an options bag with named `includeHistorical` property.

warner replied:
ok, I'll do that, but FWIW the only call site is internal, in `importer.js`

gibson042 commented on packages/swing-store/src/snapStore.js:
```diff
@@ -448,62 +482,122 @@ export function makeSnapStore(
    * @param {string} name  Artifact name of the snapshot
-   * @param {SwingStoreExporter} exporter  Whence to get the bits
-   * @param {object} info  Metadata describing the artifact
+   * @param {() => AnyIterableIterator<Uint8Array>} chunkProvider  get snapshot bytes
```
```suggestion
   * @param {() => AnyIterableIterator<Uint8Array>} makeChunkIterator  get an iterator of snapshot byte chunks
```

gibson042 commented on packages/swing-store/src/snapStore.js:
```diff
@@ -448,62 +482,122 @@ export function makeSnapStore(
-    // prettier-ignore
-    info.snapPos === snapPos ||
-      Fail`snapshot name says snapPos ${q(snapPos)}, metadata says ${q(info.snapPos)}`;
+    const metadata = sqlGetSnapshotHashFor.get(vatID, snapPos);
+    metadata || Fail`no metadata for snapshot ${name}`;
```
Potentially:
```suggestion
    const metadata = sqlGetSnapshotHashFor.get(vatID, snapPos) ||
      Fail`no metadata for snapshot ${name}`;
```

gibson042 commented on packages/swing-store/src/snapStore.js:
```diff
@@ -448,62 +482,122 @@ export function makeSnapStore(
+    assert.equal(level, 'operational'); // for now
+    // every 'inUse' snapshot must be populated
+    const vatIDs = sqlListPrunedCurrentSnapshots.all();
+    if (vatIDs.length) {
+      throw Fail`current snapshots are pruned for vats ${vatIDs.join(',')}`;
+    }
```
```suggestion
    vatIDs.length === 0 || Fail`current snapshots are pruned for vats ${vatIDs.join(',')}`;
```

warner replied:
I know this pattern is in vogue around here, but I don't really like it. It's shorter, and I like shorter, but it takes longer for me to think about the logic. Simpler is better. Things that read more like English are better.

gibson042 commented on packages/swing-store/src/transcriptStore.js:
```diff
@@ -516,55 +542,131 @@ export function makeTranscriptStore(
-   * @param {SwingStoreExporter} exporter  Exporter from which to get the span data
-   * @param {object} info  Metadata describing the span
+   * @param {() => AnyIterableIterator<Uint8Array>} chunkProvider  get snapshot bytes
+   * @param {boolean} includeHistorical
    *
    * @returns {Promise<void>}
    */
-  async function importSpan(name, exporter, info) {
+  async function populateTranscriptSpan(
+    name,
+    chunkProvider,
+    includeHistorical,
+  ) {
```
Same applies here (naming of `chunkProvider` and options vs. boolean trap `includeHistorical`).

gibson042 commented on packages/swing-store/src/transcriptStore.js:
```diff
@@ -516,55 +542,131 @@ export function makeTranscriptStore(
-      Fail`artifact name says endPos ${q(endPos)}, metadata says ${q(info.endPos)}`;
-    const artifactChunks = exporter.getArtifact(name);
+
+    const metadata = sqlGetSpanMetadataFor.get(vatID, startPos, endPos);
+    metadata || Fail`no metadata for transcript span ${name}`;
```
```suggestion
    const metadata = sqlGetSpanMetadataFor.get(vatID, startPos, endPos) ||
      Fail`no metadata for transcript span ${name}`;
```

gibson042 commented on packages/swing-store/test/util.js:
```diff
@@ -0,0 +1,26 @@
+ * @param {string} [prefix]
+ * @returns {Promise<[string, () => void]>}
+ */
+export const tmpDir = prefix =>
```
It looks like we've written this function a few times. Can you move it to `@agoric/internal` and maybe TODO de-duplication?

mhofman replied:
They're all slightly different. I don't think we should abstract this in this PR. Maybe open an issue instead?

gibson042 replied:
#8109

gibson042 reviewed (commented):
Thanks for the SwingStore documentation, and for the tests! I think the overwhelming majority of my comments relate to stylistic choices, and the core of it looks good. But I am inclined to agree with @mhofman about error messages.

warner commented on packages/swing-store/docs/swingstore.md:
```diff
@@ -0,0 +1,52 @@
+
+The state is broken up into several pieces, or "stores":
+
+* `bundleStore`: a string-keyed Bundle-value table, holding source bundles which can be evaluated by `importBundle` to create vats, or new Compartments within a vat
```
they are property names on the `kernelStorage` object, so yeah I think "`bundleStore`" is better than "**bundleStore**"

warner commented on packages/swing-store/docs/swingstore.md:
```diff
@@ -0,0 +1,52 @@
+* `bundleStore`: a string-keyed Bundle-value table, holding source bundles which can be evaluated by `importBundle` to create vats, or new Compartments within a vat
+* `transcriptStore`: records a linear sequence of deliveries and syscalls (with results), collectively known as "transcript entires", for each vat
+* `snapStore`: records one or more XS heap snapshots for each vat, to rebuild a worker more efficiently than replaying all transcript entries from the beginning
+* `kvStore`: a string-keyed string-valued table, which holds everything else. Currently, this holds each vat's c-list and vatstore data, as well as the kernel-wide object and promise tables, and run-queues.
```
Hm, I don't think so.. the `vatstore` is a database made available to all vats, via the four syscalls. That document explains how liveslots (as opposed to any other vat runtime) chooses to use the vatstore (which is an internal concern of liveslots, not something the swingstore should know anything about).

If we were to link it to anything, it should be to something that explains what the vatstore is good for.. some kind of "here's the environment that the kernel provides to all vats" document. I did a quick skim through https://github.com/Agoric/agoric-sdk/tree/master/packages/SwingSet/docs but we don't really have anything suitable there. https://github.com/Agoric/agoric-sdk/tree/master/packages/SwingSet/docs/vat-environment.md would be a candidate except that it focuses on the JS environment, rather than what's available on `syscall`.

Some day we might add a `syscalls.js` to that directory, at which point the section on the vatstore would be a good target. For now I'm going to leave it up to the reader to grep around for what a vatstore is.

warner commented on packages/swing-store/docs/swingstore.md:
```diff
@@ -0,0 +1,52 @@
+
+## Incarnations, Spans, Snapshots
+
+The kernel tracks the state of one or more vats. Each vat's execution is split into "incarnations", which are separated by a "vat upgrade" (a call to `E(vatAdminFacet).upgrade(newBundleCap, options)`). Each incarnation gets a new worker, which erases the heap state and only retains durable vatstore data across the upgrade. Every active vat has a "current incarnation", and zero or more "historic i ...
```
https://github.com/Agoric/agoric-sdk/blob/master/packages/SwingSet/docs/vat-upgrade.md is probably where they should be documented, but it was written before we started using the word "incarnation" more rigorously. I'll add a link.

warner commented on packages/swing-store/docs/swingstore.md:
```diff
@@ -0,0 +1,52 @@
+
+The kernel tracks the state of one or more vats. Each vat's execution is split into "incarnations", which are separated by a "vat upgrade" (a call to `E(vatAdminFacet).upgrade(newBundleCap, options)`). Each incarnation gets a new worker, which erases the heap state and only retains durable vatstore data across the upgrade. Every active vat has a "current incarnation", and zero or more "historic i ...
+
+Within each incarnation, execution is broken into one or more "spans", with a "current span" and zero or more "historic spans". This breaks up the transcript into corresponding spans.
+
+Each historic span ends with a `save-snapshot` entry which records the creation and saving of an XS heap snapshot. The initial span starts with a `start-worker` entry, while all non-initial spans start with a `load-snapshot` entry. The final span of historic incarnations each end with a `shutdown-worker` entry.
```
I want to emphasize the *execution* being broken up as the primary thing, and the transcripts being split into spans being a secondary consequence. The `save-snapshot`/etc entries are useful for a human trying to read the spans, or trying to understand why things are broken up into spans, but the swingstore should not pay attention to the entry contents, so I don't want these docs to lead with them.

warner commented on packages/swing-store/docs/swingstore.md:
```diff
@@ -0,0 +1,52 @@
+
+Each `save-snapshot` entry adds a new snapshot to the `snapStore`, so each vat has zero or more snapshots, of which the last one is called the "current" or "in-use" snapshot, and the earlier ones are called "historical snapshots".
+
+(note: the `deliveryNum` counter is scoped to the vat and does not reset at incarnation or span boundaries)
```
not really, I just checked and it doesn't appear in `SwingSet/docs/`

warner commented on packages/swing-store/docs/swingstore.md:
```diff
@@ -0,0 +1,52 @@
+
+The import/export process (using `makeSwingStoreExporter` and `importSwingStore`) defines some number of "artifacts" to contain much of the SwingStore data. Each bundle is a separate artifact, as is each heap snapshot. Each transcript span is a separate artifact (an aggregate of the individual transcript entries comprising that span).
+
+During export, the `getArtifactNames()` method provides a list of all available artifacts, while `getArtifact(name)` is used to retrieve the actual data. The import function processes each artifact separately.
```
Eh, the names are owned by the swingstore implementation: callers should not try to interpret them. So any such docs would be marked as internal.

warner commented on packages/swing-store/docs/swingstore.md:
```diff
@@ -0,0 +1,52 @@
+
+For this reason, the swingstore must always contain the current transcript span, and the current snapshot (if any), for every vat.
+
+However, to save space, historical spans/snapshots might be pruned, by deleting them from the database. Historical snapshots are pruned by default (unless `openSwingStore()` is given an options bag with `keepSnapshots: true`). Historical spans are not currently pruned (the `keepTranscripts` option defaults to `true`), but that may change.
```
yeah, that seems like a good improvement.. I'm trying get to establish jargon for three states for each thing: "present", "pruned", and "deleted", but when explaining what "pruned" means, we do need to say that *something* has been deleted, and deleting the "contents" or "items" is a good term to use

warner commented on packages/swing-store/docs/swingstore.md:
```diff
@@ -0,0 +1,52 @@
+* the export-time options might have done the same
+* the original DB was itself already pruned, so the data was not available in the first place
+
+In the future, a separate SwingStore API will exist to allow previously-pruned artifacts to be repopulated. Every artifact has a metadata record which *is* included in the export (in the `exportData` section, but separate from the kvStore shadow table entries), regardless of pruning modes, to ensure that this API can check the integrity of these repopulated artifacts. This reduces the reliance se ...
```
yeah, `./data-export.md` covers it

warner commented on packages/swing-store/src/bundleStore.js:
```diff
@@ -62,20 +66,36 @@ export function makeBundleStore(db, ensureTxn, noteExport = () => {}) {
-    VALUES (?, ?)
+  // the PRIMARY KEY constraint requires the bundleID not already
+  // exist
+  const sqlAddBundleRecord = db.prepare(`
+INSERT INTO bundles (bundleID, bundle) VALUES (?, NULL)
   `);
```
my editor keeps deleting the whitespace in those quasiliteral blocks..

warner commented on packages/swing-store/src/bundleStore.js:
```diff
@@ -62,20 +66,36 @@ export function makeBundleStore(db, ensureTxn, noteExport = () => {}) {
+
+  function populateBundle(bundleID, serialized) {
     ensureTxn();
+    sqlHasBundleRecord.run(bundleID) || Fail`missing ${bundleID}`;
+    sqlPopulateBundleRecord.run({ bundleID, serialized });
```
Nope, the first statement is asserting that we already have the metadata present. If we combined them, then even with the `changes > 0` check, we'd be allowing the repopulation to invent new metadata that wasn't previously established. That would allow the set of artifacts to create a new truth, when repopulation is supposed to be limited to previously-accepted truths.

Changing `addBundle` to do everything in a single step would be ok, but the call from `importBundle` would not (even during an import, we treat the export-data as truth, and the artifact/artifact-names as untrusted). And when we later implement `repopulateBundle`, the issue will really stand out.

Also, I don't really want to rely upon the SQLite binding's behavior so much.. a casual reader can guess that `run()` takes a statement, but it's not very obvious what a `{ changes }` return value might mean.

warner commented on packages/swing-store/src/bundleStore.js:
```diff
@@ -172,8 +229,10 @@ export function makeBundleStore(db, ensureTxn, noteExport = () => {}) {
       Fail`expected artifact name of the form 'bundle.{bundleID}', saw ${q(name)}`;
-    const rawBundle = sqlGetBundle.get(bundleID);
-    rawBundle || Fail`bundle ${q(name)} not available`;
+    const row = sqlGetBundle.get(bundleID);
+    row || Fail`bundle ${q(bundleID)} not found`;
+    const rawBundle = row.bundle;
+    rawBundle || Fail`bundle ${q(bundleID)} pruned`;
```
I'll apply the change, but I don't think a two-line function is worth factoring out.

warner commented on packages/swing-store/src/exporter.js:
```diff
@@ -0,0 +1,169 @@
+ *
+ * Get a list of name of artifacts available from the swingStore.  A name returned
+ * by this method guarantees that a call to `getArtifact` on the same exporter
+ * instance will succeed. Options control the filtering of the artifact names
+ * yielded.
```
The options are on `makeSwingStoreExporter` and remembered internally.. I'll update the comment

warner commented on packages/swing-store/src/exporter.js:
```diff
@@ -0,0 +1,169 @@
+
+/**
+ * @param {string} dirPath
+ * @param {string} exportMode
```
Good idea.

We don't.. I'll add text to data-export.md to explain the available modes.

warner commented on packages/swing-store/src/exporter.js:
```diff
@@ -0,0 +1,169 @@
+ * @returns {SwingStoreExporter}
+ */
+export function makeSwingStoreExporter(dirPath, exportMode = 'current') {
+  typeof dirPath === 'string' || Fail`dirPath must be a string`;
```
Hm, if I had cause to change anything else about this area, I'd probably change it to `assert.typeof` simply out of brevity. but I don't, so I'm happy to leave it using `|| Fail`. The real purpose is to catch simple mistakes at runtime (assuming that `tsc` or an IDE didn't catch them earlier).

warner commented on packages/swing-store/src/exporter.js:
```diff
@@ -0,0 +1,169 @@
+   * @yields {string}
+   */
+  async function* getArtifactNames() {
+    yield* snapStore.getArtifactNames(exportHistoricalSnapshots);
+    yield* transcriptStore.getArtifactNames(exportHistoricalTranscripts);
+    yield* bundleStore.getArtifactNames();
```
yeah, right? makes it all very clean and easy to read

warner commented on packages/swing-store/src/exporter.js:
```diff
@@ -0,0 +1,169 @@
+    }
+  }
+
+  const sqlAbort = db.prepare('ROLLBACK');
```
Yeah, I think @FUDCo wanted to define them next to their point of use, rather than all in one block, but I'm good with moving `sqlAbort` higher up. I'd prefer to keep `close()` at the end, so that the functions are defined in the same order in which they'll be invoked.

warner replied:
actually, on second though that looks kind of awkward, but I'll add a comment near the `sqlBeginTransaction.run()` to explain what we're planning to do with this txn

warner commented on packages/swing-store/src/importer.js:
```diff
@@ -0,0 +1,148 @@
+import { makeSwingStore } from './swingStore.js';
+import { buffer } from './util.js';
+
+export function assertComplete(internal, level) {
```
we don't, but I'll add it to data-export.md along with https://github.com/Agoric/agoric-sdk/pull/8075#discussion_r1277930487

warner commented on packages/swing-store/src/importer.js:
```diff
@@ -0,0 +1,148 @@
+
+  // For every exportData entry, we add a DB record. 'kv' entries are
+  // the "kvStore shadow table", and are not associated with any
+  // artifacts. All other entries are associated with an artifact,
+  // however the import may or may not contain that artifact (the
+  // dataset can be incomplete: either the original DB was pruned at
+  // some point, or the exporter did not choose to include
+  // everything). The DB records we add are marked as incomplete (as
+  // if they had been pruned locally), and can be populated later when
+  // the artifact is retrieved.
```
In this case, the dataset actually lacks the artifact: `getArtifactNames()` will omit it, and `getArtifact(name)` will throw. We can talk about "records lacking data" (so "records" are tri-state: present, pruned, or absent), but an "artifact" is much more binary (present in both `getArtifactNames` and `getArtifact`, or missing from both).

warner commented on packages/swing-store/src/importer.js:
```diff
@@ -0,0 +1,148 @@
+ * @returns {Promise<import('./swingStore').SwingStore>}
+ */
+export async function importSwingStore(exporter, dirPath = null, options = {}) {
+  if (dirPath) {
+    typeof dirPath === 'string' || Fail`dirPath must be a string`;
+  }
```
actually I think I'll go with
```js
  if (dirPath && typeof dirPath !== 'string') {
    Fail`dirPath must be a string`;
  }
```
which is more clear about what the requirement is

warner commented on packages/swing-store/src/importer.js:
```diff
@@ -0,0 +1,148 @@
+  const allMetadata = new Map();
+
+  for await (const [key, value] of exporter.getExportData()) {
+    const [tag] = key.split('.', 1);
+    if (tag === 'kv') {
+      // 'kv' keys contain individual kvStore entries
+      const subKey = key.substring(tag.length + 1);
```
I know it's a bit of a waste, but I think I prefer the parallelism with the subsequent loops: they can all do the same `const [tag] = key.split('.', 1)`.

warner commented on packages/swing-store/src/importer.js:
```diff
@@ -0,0 +1,148 @@
+      const metadata = JSON.parse(value);
+      if (!!key.endsWith('.current') !== !!metadata.isCurrent) {
+        throw Fail`transcript key ${key} mismatches metadata ${metadata}`;
+      }
```
oh, thanks for the reminder about `Boolean`. I think I'll prefer
```js
if (key.endsWith('.current') !== Boolean(metadata.isCurrent)) {
  Fail..
}
```

warner commented on packages/swing-store/src/importer.js:
```diff
@@ -0,0 +1,148 @@
+  // that check if they want.
+
+  for await (const name of exporter.getArtifactNames()) {
+    const chunkProvider = () => exporter.getArtifact(name);
+    const dataProvider = async () => buffer(Readable.from(chunkProvider()));
```
Agreed that `chunkProvider` isn't super clear. But I'll miss the parallelism between XProvider and YProvider (for chunks/data). The `dataProvider` could become.. `makeData`? Doesn't seem quite right.

I'll change it as you said and maybe some better name for `dataProvider` will come to me.

warner replied:
looks like the Readable.from isn't necessary here, but is still necessary in snapStore/bundleStore where I call `.pipe()` on it

warner commented on packages/swing-store/src/importer.js:
```diff
@@ -0,0 +1,148 @@
+ *
+ * @param {import('./exporter').SwingStoreExporter} exporter
+ * @param {string | null} [dirPath]
+ * @param {object} options
```
yeah, I'll add a type, the only field is `includeHistorical`

warner commented on packages/swing-store/src/importer.js:
```diff
@@ -47,158 +40,109 @@ export async function importSwingStore(exporter, dirPath = null, options = {}) {
+    } else if (tag === 'transcript') {
+      const metadata = JSON.parse(value);
+      if (!!key.endsWith('.current') !== !!metadata.isCurrent) {
+        throw Fail`transcript key ${key} mismatches metadata ${metadata}`;
```
I've added more checks

warner commented:
@mhofman I think I addressed all your feedback, except for questions about:
* should `repairMetadata()` do an internal commit (currently yes), and/or should it return a SwingStore (currently no)
* should `repairMetadata()` enforce correctness of non-current metadata (currently yes)

@gibson042 I think I addressed all of your feedback (although I did not perform all suggested edits)

mhofman commented on packages/swing-store/src/snapStore.js:
```diff
@@ -503,11 +526,25 @@ export function makeSnapStore(
+    const [tag, keyVatID, keySnapPos] = key.split('.');
+    assert.equal(tag, 'snapshot');
+    if (keySnapPos === 'current') {
+      // "snapshot.${vatID}.current" entires are meta-metadata: they
```
```suggestion
      // "snapshot.${vatID}.current" entries are meta-metadata: they
```

mhofman commented on packages/swing-store/test/test-export.js:
```diff
@@ -63,7 +64,7 @@ const exportTest = test.macro(async (t, mode) => {
 
   assert.typeof(mode, 'string');
-  let exportMode = mode;
+  let exportMode = /** @type { 'current' | 'archival' | 'debug' } */ (mode);
```
Import the `@typedef` here instead of redefining the type

mhofman reviewed (approved):
I really like the new layering!

Thanks also for adding back some consistency checks. There are still a few places where metadata key shapes are not checked (extra pieces ignored silently), but that's not super important.

Same with the other comments, all small nits.

warner commented on packages/swing-store/src/bundleStore.js:
```diff
@@ -210,6 +209,16 @@ INSERT INTO bundles (bundleID, bundle) VALUES (?, NULL)
+  function importBundleRecord(key, value) {
+    const [tag, ...pieces] = key.split('.');
+    assert.equal(tag, 'bundle');
+    const bundleID = pieces[0];
+    assert.equal(bundleID, value);
```
except it needs to be `assert.equal(pieces.length, 1)`

warner commented on packages/swing-store/src/bundleStore.js:
```diff
@@ -229,10 +238,9 @@ INSERT INTO bundles (bundleID, bundle) VALUES (?, NULL)
-    rawBundle || Fail`bundle ${q(bundleID)} pruned`;
+    const row =
+      sqlGetBundle.get(bundleID) || Fail`bundle ${q(bundleID)} not found`;
+    const rawBundle = row.bundle || Fail`bundle ${q(bundleID)} pruned`;
```
So far I've found that I don't care for `good_condition || Fail..`, because it makes me have to think about what's being short-circuited (it's not immediately obvious that the left-hand expression is a good thing, rather than a bad thing). But this form doesn't bug me in that same way, maybe because it reads more like english, or maybe the fact that it needs to evaluate to something makes it more clear that the bad_condition would be if the left expression were falsy. Eh.

warner commented on packages/swing-store/src/bundleStore.js:
```diff
@@ -277,7 +285,11 @@ INSERT INTO bundles (bundleID, bundle) VALUES (?, NULL)
     await 0; // no synchronous prefix
-    const [_tag, bundleID] = name.split('.', 2);
+    const parts = name.split('.');
+    if (parts[0] !== 'bundle' || parts.length !== 2) {
+      throw Fail`expected name like "bundle.{bundleID}", got ${q(name)}`;
+    }
```
Yeah, good idea:
```js
function bundleIDFromName(name) {
  typeof name === 'string' || Fail`artifact name must be a string`;
  const [tag, ...pieces] = key.split('.');
  if (tag !== 'bundle' || pieces.length !== 1) {
    Fail`expected artifact name of the form 'bundle.{bundleID}', saw ${q(name)}`;
  }
  const bundleID = pieces[0];
  return bundleID;
}
```

warner commented on packages/swing-store/src/exporter.js:
```diff
@@ -74,7 +76,7 @@ import { makeTranscriptStore } from './transcriptStore.js';
 /**
  * @param {string} dirPath
- * @param {string} exportMode
+ * @param {'current' | 'archival' | 'debug'} exportMode
```
Sure thing.. exporter.js will have:
```js
/**
 * @typedef {'current' | 'archival' | 'debug'} ExportMode
 */

/**
 * @param {string} dirPath
 * @param { ExportMode } exportMode
 * @returns {SwingStoreExporter}
 */
export function makeSwingStoreExporter(dirPath, exportMode = 'current') {
```

and `types.d.ts` will have
```js
export type { SwingStoreExporter, ExportMode } from './src/exporter.js';
```

warner commented on packages/swing-store/src/importer.js:
```diff
@@ -114,24 +110,22 @@ export async function importSwingStore(exporter, dirPath = null, options = {}) {
-        includeHistorical,
+        makeChunkIterator,
+        { includeHistorical },
       );
```
yeah, good idea, I'll add a comment to remind us of that cleanup opportunity

warner commented on packages/swing-store/src/repair-metadata.js:
```diff
@@ -0,0 +1,76 @@
+  for (const [key, value] of allMetadata.entries()) {
+    const [tag] = key.split('.', 1);
+    if (tag === 'bundle') {
+      // we ignore bundle records, because we're also ignoring
+      // artifacts, so any new bundle records stand no chance of being
+      // populated, which would fail our completeness check. Also,
+      // #8025 was accidentally deleting historical
+      // snapshots/transcript records, not bundles.
```
Hm, yeah, that seems slighly safer, although I'm not sure what name to use. `bundleStore.repairBundleRecord` wouldn't be right: that would imply we're repopulating/replacing something missing. The function would need to return silently if the record were already present, and throw if it were missing, which is more like "assertBundleRecordPresent". (part of the reason I skipped this case was because using the right name would break up the symmetry of the other `repairXRecord` calls)

Hm, or, I suppose I *could* call it `repairBundleRecord`, but then in bundleStore.js declare that their lack of metadata and their required presence for `operational` mode means that there is no repair, leaving just the "ignore if already present" feature in place, and throw otherwise. Yeah, I'll do that.

warner commented on packages/swing-store/src/snapStore.js:
```diff
@@ -483,10 +497,19 @@ export function makeSnapStore(
+      // metadata['snapshot.v1.current'] = 'snapshot.v1.5' , i.e. it
+      // points to the name of the current artifact. We could
+      // conceivably remember this and compare it against the .inUse
+      // property of that record, but it's not worth the effort (we
+      // might encounter the records in either order).
```
Yeah, my comment was just meant to explain why the effort was nontrivial: creating that Map, tracking the possible states, scanning all observed vats at the end to see if the `.current` was missing (or did we need one anyways? which means tracking transcript spans and incarnations), etc.

gibson042 commented on packages/swing-store/src/swingStore.js:
```diff
@@ -483,6 +477,12 @@ export function makeSwingStore(dirPath, forceReset, options = {}) {
     stopTrace();
   }
 
+  const internal = makeInternal(snapStore, transcriptStore, bundleStore);
```
Man, `makeInternal` really wants a better name. Maybe this?
```suggestion
  const internal = makeSwingStoreKit(snapStore, transcriptStore, bundleStore);
```
Or alternatively just don't bother with the function at all:
```suggestion
  /** @type {SwingStoreInternal} */
  const internal = harden({ snapStore, transcriptStore, bundleStore });
```

warner replied:
I like the latter, but I need a place to put `SwingStoreInternal` where it can be imported without causing a cycle. I can create a new file to hold just the type (no actual JS code). Should it be called `internal.d.ts`? Do I need to use a `@type {import('./internal.d.ts').SwingStoreInternal}` to reach it?

warner replied:
ah, I see, just `internal.js` with an `export {};`, like in the new types.js
